docs(api): Describe Flex-style deck slot names in the Python Protocol API

SyntaxColoring opened the pull request:
# Overview

Update https://docs.opentrons.com for the Flex's new style of deck slot names, like `"A1"`. Closes RTC-286.

# Changelog

* Update the reference documentation for any method that takes a deck slot as a parameter, or returns a deck slot.
* Add a new top-level "Deck Slots" page. This is a central place to introduce the two formats, and to describe their interchangeability. The reference documentation links to it.

And an unrelated thing that I'm squeezing into this PR:

* Make sure the `OFF_DECK` constant will get documented and cross-referenced properly in the future, when we publicize `move_labware()`. This takes a bit of Sphinx trickery.

# Review requests

Preview on http://sandbox.docs.opentrons.com/deck_label_docs/v2/.

* This is intended to be a full, complete, comprehensive update of every place in the Python Protocol API that deals with deck slots. I worked off of @jbleon95's [list](https://opentrons.atlassian.net/wiki/spaces/RPDO/pages/3708256298/Audit+of+deck+slots+in+the+Python+Protocol+API+and+Protocol+Engine). Please let me know if it looks like we missed any place.
* Do we agree with adding the new "Deck Slots" page? Do we agree with its position in the table of contents?

# Risk assessment

Low.

SyntaxColoring committed:
OFF_DECK tangent.

SyntaxColoring committed:
Update deck slot docs.

SyntaxColoring commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -777,19 +791,23 @@ def location_cache(self, loc: Optional[Location]) -> None:
     @property  # type: ignore
     @requires_version(2, 0)
     def deck(self) -> Deck:
```
One thing I'm omitting here is that if you iterate over the keys, like `for deck_slot_name in protocol.deck`, the format will change depending on which robot the protocol is for. I couldn't figure out how to describe that concisely. Any suggestions?

SyntaxColoring replied:
It sort of feels like we're getting to the point where the `Deck` class and its methods deserve their own reference documentation, instead of trying to stuff it all into the docstring of `ProtocolContext.deck`.

ecormany replied:
This would be great for making the new Deck Slots page less stubby. Can be another commit to this PR or as a follow-up.

sanni-t commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -777,19 +791,23 @@ def location_cache(self, loc: Optional[Location]) -> None:
         and :py:attr:`loaded_modules` to see a dict of modules.
 
-        For advanced control you can delete an item of labware from the deck
+        For advanced control, you can delete an item of labware from the deck
         with e.g. ``del protocol.deck['1']`` to free a slot for new labware.
```
While we're at it, I think we should mention that deck item deletion is not supported from 2.14 onwards. 

If I tested this correctly, on API v2.14+, deleting deck items results in TypeError saying that deck item deletion is not supported. I have not gotten to the bottom of why we get this typeerror, but even without the type error, deck item deletion doesn't work with engine backed API.

ecormany replied:
I think the best advice would be to direct users to `move_labware` off deck in higher API versions. (That also should work in 2.14 unless feature flags have changed recently…)

SyntaxColoring replied:
Starting a discussion in `#proj-python-protocol-api`. It seems like we'd want to make `del deck[foo]` an alias of `protocol.moveLabware(foo, OFF_DECK)`?

sanni-t reviewed (approved):
LGTM! Just the one comment about deck item deletion.
And I agree about making a separate deck documentation

ecormany commented on api/docs/v2/deck_slots.rst:
```diff
@@ -0,0 +1,74 @@
+
+When you load an item onto the robot's deck, like with :py:obj:`ProtocolContext.load_labware()` or :py:obj:`ProtocolContext.load_module()`, you need to specify which slot to put it in.
+
+When doing that, you can choose between two formats:
```
```suggestion
Specify a slot in one of two formats:
```

ecormany commented on api/docs/v2/deck_slots.rst:
```diff
@@ -0,0 +1,74 @@
+
+When doing that, you can choose between two formats:
+
+* A letter-number pair like ``"A1"``. This matches how the deck is physically labeled on an Opentrons Flex.
+* A number like ``"10"`` or ``10``. This matches how the deck is physically labeled on an Opentrons OT-2.
```
Maybe easier to parse as a table, like:
```suggestion
+----------------+-------------+----------------+
| Robot          | Slot format | Example        |
+================+=============+================+
| Opentrons Flex | Numeric     | `"A1"`         |
+----------------+-------------+----------------+
| Opentrons OT-2 | Coordinate  | `"10"` or `10` |
+----------------+-------------+----------------+
```

SyntaxColoring replied:
Leaving this out for now because I liked explicitly spelling out "this matches how the deck is physically labeled on an Opentrons {Flex,OT-2}". I thought it would be helpful context if you've only ever seen one of these robots, and you're wondering WTF I'm talking about with these \~two styles of deck slot names\~.

Happy for this to be rewritten later if anyone disagrees, though.

ecormany commented on api/docs/v2/deck_slots.rst:
```diff
@@ -0,0 +1,74 @@
+==========================
+
+..
+    TODO(mm, 2023-06-05): Embed a nice SVG instead of having these tables.
```
Should have one of these very soon. Can then remove the large headers and put some transition text between the table and the image.

ecormany commented on api/docs/v2/deck_slots.rst:
```diff
@@ -0,0 +1,74 @@
+
+The Flex and OT-2 formats are interchangeable. You can use either format, regardless of which robot your protocol is for.
+
+For example, these are equivalent:
```
We might want to be clear on whether you can mix and match within protocols, or whether you have to pick one system and stick to it (even if it doesn't match your requirements block). The fact that these are put in separate snippets is perhaps too subtle. If they were in the same protocol, would analysis raise an error on the second one, since the slot is already occupied?

SyntaxColoring replied:
Leaving this out because I couldn't quite figure out how to phrase it, but I think it's a good idea and we should do it.

ecormany commented on api/docs/v2/index.rst:
```diff
@@ -11,6 +11,7 @@ Welcome
     versioning
     new_labware
     new_modules
+    deck_slots
```
I think this is fine for :shipit: purposes — these changes won't be deployed for a bit — but I'm still thinking about how the TOC and information architecture will look in a couple months… 🤔

ecormany commented on api/docs/v2/new_protocol_api.rst:
```diff
@@ -71,6 +74,10 @@ Useful Types and Definitions
+..
+   TODO(mm, 2023-06-06): This should be uncommented when move_labware() is published. See above.
+   .. autodata:: opentrons.protocol_api.OFF_DECK
+      :no-value:
```
🎟️ RTC-288

ecormany commented on api/src/opentrons/protocol_api/_types.py:
```diff
@@ -8,3 +8,11 @@ class OffDeckType(enum.Enum):
+# Set __doc__ manually as a workaround. When this docstring is written the normal way, right after
+# the constant definition, Sphinx has trouble picking it up.
+OFF_DECK.__doc__ = """\
+A special placeholder value for a labware's location, indicating that it's not currently on the robot's deck.
```
```suggestion
A special location value, indicating that a labware is not currently on the robot's deck.
```
"placeholder" has a connotation of being _default_, which this definitely isn't.

SyntaxColoring replied:
Good note, thanks.. Just changing "the labware" to "a labware."

ecormany commented on api/src/opentrons/protocol_api/labware.py:
```diff
@@ -342,11 +342,16 @@ def uri(self) -> str:
-        If on a module, the parent :py:class:`ModuleContext` will be returned.
-        If off deck, `None` will be returned.
+        Returns:
+            If the labware is directly on the robot's deck, the `str` name of the deck slot.
+            On a Flex, this will be like ``"D1"``, and on an OT-2, this will be like ``"1"``.
+            See :ref:`deck-slots`.
```
With the addition of the new page, I'd be cool with:
```suggestion
            If the labware is directly on the robot's deck, the `str` name of the :ref:`deck slot <deck-slots>`.
```

And similarly below, link rather than spelling out in docstrings.

SyntaxColoring replied:
I think I'd prefer to keep these as-is because, at the time of writing, we're not consistent with *returned* deck slots, so it doesn't seem right to describe "general behavior" somewhere and then link to that description from everywhere.

`Protocol.loaded_modules` and `Protocol.loaded_labwares` return integers even on Flexen, but these methods return coordinate strings. Since it's currently dependent on the specific method or property you're calling, it seems appropriate to describe in the method reference.

ecormany commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -500,15 +506,21 @@ def load_module(
+
+            Some modules, like the Thermocycler, are only valid in one deck location.
+            You do not have to specify a location when loading a Thermocycler---it will always
+            be in Slot 7.
```
Is it always treated as B1 on Flex then? Maybe we can just eliminate the en dash and everything after and _not worry about it_. That's what we want users to do!

SyntaxColoring replied:
It is treated as B1 on the Flex, yeah. Updated to mention this.

ecormany commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -500,15 +506,21 @@ def load_module(
+            be in Slot 7.
+
+            .. versionchanged:: 2.15
+                You can now specify a deck slot as a letter-number pair, like ``"D1"``.
```
```suggestion
                You can now specify a deck slot with coordinates, like ``"D1"``.
```
And do we need a similar notice on `load_labware`?

SyntaxColoring replied:
Yes, thank you.

ecormany commented on api/src/opentrons/protocol_api/labware.py:
```diff
@@ -342,11 +342,16 @@ def uri(self) -> str:
+            On a Flex, this will be like ``"D1"``, and on an OT-2, this will be like ``"1"``.
+            See :ref:`deck-slots`.
+
+            If the labware is on a module, a :py:class:`ModuleContext`.
```
NB: this makes nice formatting even though there's no referenceable thing that is a `ModuleContext` (each type of module has its own context class).

ecormany reviewed (approved):
This is a great start to documenting this set of features — thank you, Max! I left many comments, some of which we'll probably want to act upon before merging, but no big pressure on any of them. All of this will be rolled up into a much larger public deployment describing Flex features later this summer. So if you're feeling [bold](https://en.wikipedia.org/wiki/Wikipedia:Be_bold), go ahead and get this into `edge`.

SyntaxColoring committed:
Apply suggestions from code review

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

SyntaxColoring committed:
Add ticket references.

SyntaxColoring committed:
Merge branch 'edge' into deck_label_docs

SyntaxColoring committed:
Merge branch 'edge' into deck_label_docs

Resolve docstring conflicts in:

SyntaxColoring committed:
"letter-number pair" -> "coordinate"

SyntaxColoring committed:
Add missing versionchanged note for load_labware().

SyntaxColoring committed:
Update Thermocycler slots.

Describe which Flex slots it occupies, in addition to which OT-2 slots it occupies.

Describe special slot B1 for the Flex, in addition to special slot 7 for the OT-2.

Remove specific mention of DeckConflictError, which is not a public thing that users can import. Instead, just say, vaguely, that we'll raise "an error."

SyntaxColoring committed:
Update list of edge slots.

SyntaxColoring committed:
Revert "Update list of edge slots."

This reverts commit 0b38175a9f00f6de3c2b296db983b036efa7c43d.

Generalizing optimal slots for the Flex is out of scope for this PR. I'd need to understand more about preferred venting and cable management on the Flex.

SyntaxColoring commented:
Failing tests appear to be unrelated, from `tests/opentrons/hardware_control/integration/test_thermocycler.py` flakiness.
